compiler/optimizations/inferConstRefs.cpp

```diff
@@ -366,43 +316,263 @@ static bool inferConstRef(Symbol* sym) {
+      ArgSymbol* form = actual_to_formal(use);
+      if (form->isRef()) {
+        if (!inferConstRef(form)) {
+          retval = false;
+        }
+      }
+    }
+    else if (parent && isMoveOrAssign(parent)) {
+      if (call->isPrimitive(PRIM_ADDR_OF) ||
+          call->isPrimitive(PRIM_SET_REFERENCE)) {
+        SymExpr* LHS = toSymExpr(parent->get(1));
+        if (LHS->isRef() && !inferConstRef(LHS->symbol())) {
+          retval = false;
+        }
+      }
+    }
+    else if (isMoveOrAssign(call)) {
+      if (use == call->get(1)) {
+        numDefs += 1;
+        if (numDefs > 1) {
+          retval = false;
+        }
+      }
     } else {
-      INT_ASSERT(isVarSymbol(sym));
-      sym->qual = QUAL_CONST_REF;
+      // To be safe, exit the loop with 'false' if we're unsure of how to
+      // handle a primitive.
+      retval = false;
     }
   }
 
-  // TODO: assert that it's not already finalized?
-  if (info) {
-    info->finalized = true;
+  if (isFirstCall) {
+    if (retval && !info->finalizedConstness) {
+      if (ArgSymbol* arg = toArgSymbol(sym)) {
+        INT_ASSERT(arg->intent & INTENT_FLAG_IN);
+        arg->intent = INTENT_CONST_IN;
+      } else {
+        INT_ASSERT(isVarSymbol(sym));
+        sym->qual = QUAL_CONST_VAL;
+      }
+    }
+
+    if (info) {
+      info->reset();
+      info->finalizedConstness = true;
+    }
+  } else if (!retval && info) {
+    // Not the first call, but something happened that prevents us from marking
+    // this symbol as const. Now that we know, indicate that this is finalized.
+    info->finalizedConstness = true;
+  }
+
+  return retval;
+}
+
+static bool inferRefToConst(Symbol* sym) {
+  if (!sym->isRef()) return false;
+
+  bool isConstRef = sym->qualType().getQual() == QUAL_CONST_REF;
+  const bool wasRefToConst = sym->hasFlag(FLAG_REF_TO_CONST);
+
+  ConstInfo* info = infoMap[sym];
+
+  // If this ref isn't const, then it can't point to a const thing
+  if ((info && info->finalizedRefToConst) || wasRefToConst || !isConstRef) {
+    return wasRefToConst;
+  }
+
+  bool isFirstCall = false;
+  if (info && info->alreadyCalled == false) {
+    isFirstCall = true;
+    info->alreadyCalled = true;
+  }
+
+  bool retval = true;
+
+  if (isArgSymbol(sym)) {
+    // Check each call and set retval to false if any actual is not a ref
+    // to a const.
+    FnSymbol* fn = toFnSymbol(sym->defPoint->parentSymbol);
+    if (fn->hasFlag(FLAG_VIRTUAL)) {
+      // Not sure how to best handle virtual calls, so simply set false for now
+      retval = false;
+    } else {
+      if (isFirstCall && info) {
+        INT_ASSERT(info->fnUses == NULL);
+        info->fnUses = fn->firstSymExpr();
+      } else if (info == NULL) {
+        retval = false;
+      }
+      // Need this part to be re-entrant in case of recursive functions
+      while (info && info->fnUses != NULL && retval) {
+        SymExpr* se = info->fnUses;
+        info->fnUses = se ? se->symbolSymExprsNext : NULL;
```

Comment:
Let's put that in a separate PR. It arguably will not need much testing (since it can be "trivial code movement") and we might spend a while nitpicking its name and arguments etc.